include/swift/AST/Decl.h

```diff
@@ -2373,7 +2373,13 @@ class ValueDecl : public Decl {
   /// Determines the kind of access that should be performed by a
   /// DeclRefExpr or MemberRefExpr use of this value in the specified
   /// context.
-  AccessSemantics getAccessSemanticsFromContext(const DeclContext *DC) const;
+  ///
+  /// \param DC The declaration context.
+  ///
+  /// \param base If querying for a MemberRefExpr, this is its base expression.
+  ///             Otherwise, \c nullptr.
+  AccessSemantics getAccessSemanticsFromContext(const DeclContext *DC,
+                                                Expr *base) const;
```

Comment:
Could this just be a boolean parameter instead of an Expr? Have the caller decide if its the magic 'self' decl or not.

Reply:
@slavapestov We could do, though I can't immediately think of a good name for such a parameter; calling it something like `isAccessOnSelf` makes sense for members, but not for variables in general. Any ideas?

Reply:
isSelfPropertyAccess maybe?

Reply:
@slavapestov But then we'd have to pass `true` for variables which seems a bit confusing, as we want direct accesses on both variables and members on `self`.

Reply:
@slavapestov We could always just make it an enum, but I don't know if that's a bit excessive for this one situation

Reply:
You could make it a default argument with a default value of 'true'

Reply:
@slavapestov Good idea, though to avoid confusion for future readers/when debugging, I think the parameter should be named something like `isVariableOrSelfPropertyAccess`. If that's good with you, I'll go ahead and make that change.

Reply:
I honestly like it better the way Hamish has it originally; we don't have to move the checks up to the caller and we don't end up with a parameter that jams two conditions into one. If not that, however, I think I'd prefer `isAccessOnSelf`, but you only check it if the VarDecl is a property.

Reply:
@jrose-apple Ah, good idea to check whether the var is a property within `getAccessSemanticsFromContext`, that makes it much nicer.

Honestly I'm not too attached to passing the base expression vs. a bool flag; as far as I can tell, member accesses should call through `getImplicitMemberReferenceAccessSemantics` anyway, so I don't think it's a problem having the 'self' base checking logic there.

Therefore I'll go ahead and make this change, @slavapestov could you please take a look when you get a chance?

Thank you both for your feedback!